go-controller/pkg/cni/cni.go

```diff
@@ -199,3 +179,28 @@ func (pr *PodRequest) getCNIResult(podInterfaceInfo *PodInterfaceInfo) (*current
 		IPs:        ips,
 	}, nil
 }
+
+// getAnnotationsOnPod obtains the pod annotation from the cache
+func getAnnotationsOnPod(ctx context.Context, podLister corev1listers.PodLister, namespace, name string) (map[string]string, error) {
+	var annotations map[string]string
+
+	if err := wait.PollImmediate(200*time.Millisecond, 30*time.Second, func() (bool, error) {
+		select {
+		case <-ctx.Done():
+			return false, fmt.Errorf("pod operation canceled")
+		default:
+			pod, err := podLister.Pods(namespace).Get(name)
+			if err != nil {
+				return false, err
+			}
+			annotations = pod.ObjectMeta.Annotations
+			if _, ok := annotations[util.OvnPodAnnotationName]; ok {
+				return true, nil
+			}
+		}
+		return false, nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to get pod %s/%s annotations: %v", name, namespace, err)
+	}
```

Comment:
I don't know if we should use the `wait.PollImmediate()` here and do ourselves the loop, seems clearer to me:
https://play.golang.org/p/tBJh8zn2NKQ i.e. (is an example, I have problems with indentation in comments)
```go
func getAnnotationsOnPod(ctx context.Context, podLister corev1listers.PodLister, namespace, name string) (map[string]string, error) {

var annotations map[string]string
timeout := time.After(30 * time.Second)
pollInt := 200*time.Millisecond
for {
   select {
         case <-ctx.Done():
		return nil, fmt.Errorf("pod operation canceled")
         case <- timeout:
                  return nil, fmt.Errorf("failed to get pod %s/%s annotations: %v", name, namespace, err)
          default:
                        pod, err := podLister.Pods(namespace).Get(name)
			if err != nil {
				return nil, err
			}
			annotations = pod.ObjectMeta.Annotations
			if _, ok := annotations[util.OvnPodAnnotationName]; ok {
				return annotations, nil
			}
            }
            time.Sleep(pollInt)
       }
}

Reply:
your example is what I had in mind as well

Reply:
Fixed.